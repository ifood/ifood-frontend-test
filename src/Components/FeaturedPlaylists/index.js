import React from 'react'

import { Container } from './styles.js'

export default function FeaturedPlaylist(){
    return(
        <Container>
            <h1>Ol√° mundo</h1>
        </Container>
    )
}