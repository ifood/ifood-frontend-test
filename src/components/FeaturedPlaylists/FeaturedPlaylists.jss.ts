import { makeStyles, Theme } from '@material-ui/core/styles';

export const useStyles = makeStyles(
  (theme: Theme) => ({
    indicator: {
      marginTop: theme.spacing(2),
      display: 'flex',
      justifyContent: 'center',
    },
  }),
  { name: 'FeaturedPlaylists' }
);
