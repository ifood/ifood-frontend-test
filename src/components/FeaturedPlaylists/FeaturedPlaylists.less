@import '../../styles/less/_parameters';

.playlist{
  padding:15px 0px;

  &__item{
    width: 250px;
    height: 250px;
    background-size: cover;
    background-position: center;
    cursor:pointer;

    &__info{
      width: 100%;
      height: 100%;
      padding:15px;

      display:flex;
      align-items:center;
      justify-content:center;

      background-color: @card-background;
      color:transparent;
      font-weight: 600;

      &:hover{
        background-color:@card-background-hover;
        color:@white;

      -webkit-transition: background-color 300ms linear, color 800ms linear;
      -moz-transition: background-color 300ms linear, color 800ms linear;
      -o-transition: background-color 300ms linear, color 800ms linear;
      -ms-transition: background-color 300ms linear, color 800ms linear;
      transition: background-color 300ms linear, color 800ms linear;
      }
    }
  }
}
