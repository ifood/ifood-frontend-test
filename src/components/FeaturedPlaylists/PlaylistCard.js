import React from 'react';

const PlaylistCard = (props) => (
  <div className="col-4">
    <div className="card">
    <img className="card-img-top" src={props.imageUrl} alt={props.name} />
    <div className="card-body">
      <h5 className="card-title">{props.name}</h5>
      <p className="card-text">Total tracks: {props.tracks.total}</p>
      <a href={props.externalUrl} className="btn btn-primary">See on Spotify</a>
    </div>
  </div>
  </div>
);

export default PlaylistCard;