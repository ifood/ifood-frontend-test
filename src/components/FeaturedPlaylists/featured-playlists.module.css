.noResultsText {
  flex-grow: 1;
  text-align: center;
}

.playlistWrapper {
  margin-right: 0;
}