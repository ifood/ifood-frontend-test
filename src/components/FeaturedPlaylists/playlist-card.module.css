
.cardContainer {
  margin-bottom: 15px;
  padding-right: 0;
}

.cardBody { padding: 1rem; }