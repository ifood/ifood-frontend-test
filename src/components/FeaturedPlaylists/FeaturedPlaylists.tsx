import React, { useCallback, useMemo, useState } from 'react';
import CircularProgress from '@material-ui/core/CircularProgress';
import Typography from '@material-ui/core/Typography';
import debounce from 'lodash/debounce';
import { Filter } from '../Filter/Filter';
import { Playlists } from '../Playlists/Playlists';
import { useFeaturedPlaylists, FeaturedPlaylistsParams } from '../../hooks/useFeaturedPlaylists';
import { useStyles } from './FeaturedPlaylists.jss';

export const FeaturedPlaylists: React.FC = () => {
  const classes = useStyles();
  const [search, setSearch] = useState('');
  const [params, setParams] = useState<FeaturedPlaylistsParams>({});
  const { featuredPlaylists, isLoading, isSuccess, isError } = useFeaturedPlaylists(params);

  const playlists = useMemo(() => {
    if (!featuredPlaylists) {
      return [];
    }
    if (!search) {
      return featuredPlaylists;
    }
    const normalizedSearch = search.toLowerCase();
    return featuredPlaylists.filter((featuredPlaylist) =>
      featuredPlaylist.name.toLowerCase().includes(normalizedSearch)
    );
  }, [featuredPlaylists, search]);

  const handleFilter = useCallback((filters: Record<string, string>) => {
    setParams(filters);
  }, []);

  const handleSearch = useMemo(() => {
    return debounce((text: string) => setSearch(text), 400);
  }, []);

  return (
    <>
      <Filter onFilter={handleFilter} onSearch={handleSearch} />

      {isLoading && (
        <div className={classes.indicator}>
          <CircularProgress color="primary" />
        </div>
      )}

      {isError && (
        <div className={classes.indicator}>
          <Typography>Não foi possível carregar as playlists.</Typography>
        </div>
      )}

      {isSuccess && playlists.length === 0 && (
        <div className={classes.indicator}>
          <Typography>Nenhuma playlist encontrada.</Typography>
        </div>
      )}

      {isSuccess && playlists.length > 0 && <Playlists playlists={playlists} />}
    </>
  );
};
