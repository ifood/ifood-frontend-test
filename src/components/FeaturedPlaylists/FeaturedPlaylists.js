import React from 'react';
import shortid from 'shortid';
import PlaylistCard from './PlaylistCard';
import style from './featured-playlists.module.css';

const filterByName = (name) => item =>
  item.name.toLowerCase().indexOf(name.toLowerCase()) > -1;

const FeaturedPlaylists = React.memo(({
  playlists = { items: [] },
  nameFilter = ''
}) => {
  const filteredPlaylists = nameFilter
    ? playlists.items.filter(filterByName(nameFilter))
    : playlists.items;
  return (
    <div className={`row ${style.playlistWrapper}`}>
      {
        Boolean(filteredPlaylists.length) ? filteredPlaylists.map(item =>
          <PlaylistCard
            key={shortid()}
            name={item.name}
            imageUrl={item.images[0].url}
            tracks={item.tracks}
            externalUrl={item.external_urls.spotify}
          />
        ) :
        <p className={style.noResultsText}>
          {
            nameFilter && Boolean(playlists.items.length)
              ? 'Nenhum resultado encontrado'
              : 'Selecione a playlist'
          }
        </p>
      }
    </div>
  )
});

export default FeaturedPlaylists;