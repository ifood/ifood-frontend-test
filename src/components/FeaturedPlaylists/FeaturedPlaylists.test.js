import React from 'react';
import renderer from 'react-test-renderer';
import FeaturedPlaylists from './FeaturedPlaylists';

const defaultProps = {
  playlists: {
    items: [{
      name: 'Cool Sound',
      images: [{
        url: 'some/image/path'
      }],
      tracks: {
        total: 55
      },
      external_urls: {
        spotify: 'some-external-url'
      }
    }]
  },
  nameFilter: ''
}

describe('<FeaturedPlaylists />', () => {

  it('Should render correctly', () => {
    const elem = renderer
      .create(<FeaturedPlaylists { ...defaultProps } />)
      .toJSON();
    expect(elem).toMatchSnapshot();
  });

  it(`Should render with "do you search" text`, () => {
    const elem = renderer
      .create(<FeaturedPlaylists />)
      .toJSON();
    expect(elem).toMatchSnapshot();
  });

  it(`Should render with "no results found" text`, () => {
    const elem = renderer
      .create(<FeaturedPlaylists { ...defaultProps } nameFilter="Mundo Animal" />)
      .toJSON();
    expect(elem).toMatchSnapshot();
  });

});

