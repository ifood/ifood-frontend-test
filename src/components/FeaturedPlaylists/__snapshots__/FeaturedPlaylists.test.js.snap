// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FeaturedPlaylists /> Should render correctly 1`] = `
<div
  className="row playlistWrapper"
>
  <div
    className="col-sm-6 col-md-6 col-lg-4 cardContainer"
  >
    <div
      className="card"
    >
      <img
        alt="Cool Sound"
        className="card-img-top"
        src="some/image/path"
      />
      <div
        className="card-body cardBody"
      >
        <h5
          className="card-title"
        >
          Cool Sound
        </h5>
        <p
          className="card-text"
        >
          Total tracks: 
          55
        </p>
        <a
          className="btn btn-primary"
          href="some-external-url"
        >
          See on Spotify
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`<FeaturedPlaylists /> Should render with \`do you search\` text 1`] = `
<div
  className="row playlistWrapper"
>
  <p
    className="noResultsText"
  >
    Selecione a playlist
  </p>
</div>
`;

exports[`<FeaturedPlaylists /> Should render with \`no results found\` text 1`] = `
<div
  className="row playlistWrapper"
>
  <p
    className="noResultsText"
  >
    Nenhum resultado encontrado
  </p>
</div>
`;
